libcontainer/rootfs_linux.go

```diff
@@ -1103,34 +1098,99 @@ func writeSystemProperty(key, value string) error {
 	return os.WriteFile(path.Join("/proc/sys", keyPath), []byte(value), 0o644)
 }
 
-func remount(m mountEntry, rootfs string, noMountFallback bool) error {
+func remount(m mountEntry, rootfs string) error {
+	// Flags related to atime.
+	const mntAtimeFlags = unix.MS_NOATIME | unix.MS_RELATIME |
+		unix.MS_STRICTATIME | unix.MS_NODIRATIME
+	// Flags which can be locked when inheriting mounts in a different userns.
+	// In the kernel, these are the mounts that are locked using MNT_LOCK_*.
+	const mntLockFlags = unix.MS_RDONLY | unix.MS_NODEV |
+		unix.MS_NOEXEC | unix.MS_NOSUID | mntAtimeFlags
+
 	return utils.WithProcfd(rootfs, m.Destination, func(dstFD string) error {
-		flags := uintptr(m.Flags | unix.MS_REMOUNT)
-		err := mountViaFDs(m.Source, m.srcFD, m.Destination, dstFD, m.Device, flags, "")
-		if err == nil {
+		flags := m.Flags | unix.MS_BIND | unix.MS_REMOUNT
+		// The semantics of atime on Linux are very fruity and handling this
+		// nicely is a bit of a fine art -- in particular, should runc be
+		// providing exactly what the user asked for, should we produce what
+		// the kernel would naively provide with the given arguments, or should
+		// we provide what mount(8) does?
+		//
+		// mount(8)'s behaviour is very odd (it inherits old flags in a way
+		// that makes it very annoying to unset a particular flag -- it seems
+		// to naively copy the currently-set flags which is not correct for
+		// some atime flags). If we do not pass any MS_*ATIME flags, the kernel
+		// will just inherit the host flags, which leads to inconsistent
+		// results and might not be what the user expected. It also seems
+		// inadvisable to make every bind-mount use MS_STRICTATIME ("no flags")
+		// by default, as the kernel has defaulted to relatime for a long time.
+		//
+		// So, if the user has not specified any atime flags, we use a default
+		// setting of MS_RELATIME. However, if they have explicitly cleared
+		// this flag (with our "norelatime" setting) and not specified any
+		// other flags, we default to MS_STRICTATIME since that's the only
+		// setting that makes sense here. Critically, we *always* call mount(2)
+		// with some MS_*ATIME flag set to ensure we have consistent results.
+		if m.Flags&mntAtimeFlags == 0 {
+			if m.ClearedFlags&unix.MS_RELATIME != 0 {
+				flags |= unix.MS_STRICTATIME
+			} else {
+				flags |= unix.MS_RELATIME
+			}
+			logrus.Debugf("remount(%q): add implied flag 0x%x to set=0x%x,clr=0x%x", m.Source, flags&mntAtimeFlags, m.Flags, m.ClearedFlags)
+		}
+		mountErr := mountViaFDs(m.Source, m.srcFD, m.Destination, dstFD, m.Device, uintptr(flags), "")
```

Comment:
Maybe while we're at it we can remove source and fstype arguments, as they are ignored by the kernel when the MS_REMOUNT is set. I.e. something like
```suggestion
		mountErr := mountViaFDs("",nil, m.Destination, dstFD, "", uintptr(flags), "")
```

(and the same below).

I have no strong argument in favor of doing so; my weak one is "to emphasize this is a remount".

Reply:
Will do. This is actually a change I also did in the nsexec cleanup PR. 😅